docs/proposal/kmesh_support_localityLB.md

```diff
@@ -0,0 +1,183 @@
+
+### Design Details
+
+#### configure
+ref: https://pkg.go.dev/istio.io/istio/pkg/workloadapi#LoadBalancing_Scope
+1. Configure the load balance mode (failover or strict) and scope (REGION, ZONE, etc.) of the service. This corresponds to the message LoadBalancing in workload.proto.
+```
+message LoadBalancing {
+  enum Scope {
+    UNSPECIFIED_SCOPE = 0;
+    REGION = 1;
+    ZONE = 2;
+    SUBZONE = 3;
+    NODE = 4;
+    CLUSTER = 5;
+    NETWORK = 6;
+  }
+  enum Mode {
+    UNSPECIFIED_MODE = 0;
+    STRICT = 1;
+    FAILOVER = 2;
+  }
+  repeated Scope routing_preference = 1;
+  Mode mode = 2;
+}
+```
+
+Add scope and mode information to the service map, and incorporate corresponding update logic into the control plane.
+
+`bpf/kmesh/workload/include/workload.h`
+```
+typedef struct {
+    __u32 endpoint_count;               // endpoint count of current service
+    __u32 lb_policy;                    // load balancing algorithm
+    __u32 service_port[MAX_PORT_COUNT]; // service_port[i] and target_port[i] are a pair, i starts from 0 and max value
+                                        // is MAX_PORT_COUNT-1
+    __u32 target_port[MAX_PORT_COUNT];
+    struct ip_addr wp_addr;
+    __u32 waypoint_port;
+    __u8 balance_scope[MAX_SCOPE_COUNT]; // balance scope 
+    __u8 balance_mode;                   // balance mode: Unspecified_mode 0; strict 1; failover 2
+} service_value;
+```
+
+Add corresponding fields to the `pkg/controller/workload/bpfcache/service.go`, and update logic to `pkg/controller/workload/workload_processor.go`
+
+2. Configure the locality (region, zone, subzone) and health status （HEALTHY, UNHEALTHY）of the backend. This corresponds to the message in workload.proto.
+> Although the current workload API defines seven scopes, when configuring a pod's locality, only region, zone, and subzone are configured. Therefore, matching capabilities can only be realized for these three scopes.
+```
+message Locality {
+  string region = 1;
+  string zone = 2;
+  string subzone = 3;
+}
+
+message Workload {
+    ...
+    WorkloadStatus status = 17;
+    Locality locality = 24;
+    ...
+}
+enum WorkloadStatus {
+  // Workload is healthy and ready to serve traffic.
+  HEALTHY = 0;
+  // Workload is unhealthy and NOT ready to serve traffic.
+  UNHEALTHY = 1;
+}
+```
+
+Add locality and healthStatus information to the backend map, and include corresponding update logic in the control plane.
+
+`bpf/kmesh/workload/include/workload.h`
+```
+typedef struct {
+    struct ip_addr addr;
+    __u32 service_count;
+    __u32 service[MAX_SERVICE_COUNT];
+    struct ip_addr wp_addr;
+    __u32 waypoint_port;
+    __u8 health_status; // workload_health_status_t: HEALTHY, UNHEALTHY
```

Comment:
Currently we filtered out unhealthy workload

Reply:
So, only healthy workloads are stored in bpf map by the control plane. Does locality load balance not need to concern whether the workload is healthy or not?

Reply:
We can make it simpler, even the priority set can be calculated in user space.

Reply:
Priority computation between localities occurs when a new flow is came. If priority calculation takes place at the control plane, my understanding is that we would need to precompute all possible scenarios（We are unable to perform event-driven programming that interoperates with user space, right?）, then hash different situations and store them in a BPF map. The kernel space would then query the map using source and destination locality information to obtain priority information. To simplify the problem, we could arrange combinations based on the specific values pointed to by the six routing options in scope (also including cases where only some of these match). This approach has two potential issues:
Firstly, userspace must enumerate all possible scenarios, which becomes particularly burdensome as the richness of locality information increases, leading to an exponential growth in the number of situations to be stored. Secondly, the BPF map would have to store all aforementioned scenarios, with each scenario existing in the form of the prio_map as currently designed.

Reply:
My concern mainly on : 1. ebpf instruction limit  2. dataplane sorting performance Worth a try  though